### PR TITLE
Feat/fixed symbol ids

### PR DESCRIPTION
**Pull request by lara-aigmueller**

### CODE REVIEW

File: packages/app-builder/src/angular-app-shell/files/__directory__/src/app/utils/mutate-variants.ts [hunk 1]
```diff
@@ -35,36 +35,37 @@ export function singleVariantGenerator(baseName: string, theme: string, variants
...
-  if (!variants || !Array.isArray(variants) || !variants.length) {
+/**
+ * Takes a set of properties and generates all possible property value combinations.
+ * How the property values are sorted is not relevant.
+ * @param baseName - Component base name (e.g. "button").
+ * @param theme - Theme property (e.g. "light-bg").
+ * @param properties - Array of properties.
+ * @return – An array of ...
+ */
+export function variantCombinationGenerator(baseName: string, theme: string, properties: Property[]): any[] {
```
Comment (lukasholzer) [inline packages/app-builder/src/angular-app-shell/files/__directory__/src/app/utils/mutate-variants.ts]: major: Can you create a ticket for writing a test of the variantCombinationGenerator? (or write one)– it is a pure function so it should be easy to test!

File: packages/sketch-builder/src/bin.ts [hunk 2]
```diff
@@ -36,6 +36,11 @@ export async function commandLineExecutor(): Promise<number> {
   // the meta option is optional in the main function
   let meta: Library;
 
+  // set the library version when given
```
Comment (lukasholzer) [inline packages/sketch-builder/src/bin.ts]: minor: provided? 🤔

File: packages/sketch-builder/src/element-fetcher.ts [hunk 3]
```diff
@@ -28,16 +24,35 @@ const LOCAL_RESULT_PATH = resolve('tests/fixtures/library.json');
 export class ElementFetcher {
   // private _assetHsandler: AssetHandler = new AssetHandler();
   result: (TraversedPage | TraversedLibrary)[] = [];
+  private idMapping: ObjectIdMapping;
 
   constructor(public conf: SketchBuilderConfig, public meta?: Library) {}
 
+  set objectIdMapping(value: string) {
```
Comment (lukasholzer) [inline packages/sketch-builder/src/element-fetcher.ts]: major: Can you remove the setter and getter the idMapping should be a public property from type ObjectidMapping – the parsing of a json file should be done outside this class

File: packages/sketch-builder/src/element-fetcher.ts [hunk 4]
```diff
@@ -28,16 +24,35 @@ const LOCAL_RESULT_PATH = resolve('tests/fixtures/library.json');
...
+  }
+
+  get objectIdMapping(): string {
+    return JSON.stringify(this.idMapping);
+  }
+
   async generateSketchFile(): Promise<number> {
     this.sortSymbols();
-    const drawer = new Drawer();
+    const drawer = new Drawer(this.idMapping);
```
Comment (lukasholzer) [inline packages/sketch-builder/src/element-fetcher.ts]: why you don't pass the idMapping as an argument to the `drawSymbols` method?

File: packages/sketch-builder/src/element-fetcher.ts [hunk 5]
```diff
@@ -28,16 +24,35 @@ const LOCAL_RESULT_PATH = resolve('tests/fixtures/library.json');
...
       this.conf.outFile,
     );
+
+    // set library ID when library information is given in the config file
+    if (this.conf.library !== undefined) {
+      sketch.libraryId = getLibraryId(this.idMapping);
+    }
+
     const pages = [];
     let symbolsMaster = drawer.drawSymbols({ symbols: [] } as any);
```
Comment (lukasholzer) [inline packages/sketch-builder/src/element-fetcher.ts]: here it should be passed the idMapping

File: packages/sketch-file-format/src/models/document.ts [hunk 6]
```diff
@@ -4,15 +4,16 @@ import { Page } from './page';
 
 export class Document {
   private static _instance: Document;
-  private static _objectID = UUID.generate();
   private _pages = [];
+  private _objectID: string;
 
-  constructor(pages: Page[]) {
+  constructor(pages: Page[], objectID = UUID.generate()) {
```
Comment (lukasholzer) [inline packages/sketch-file-format/src/models/document.ts]: major: don't call a function in a constructor!

File: packages/sketch-file-format/src/models/document.ts [hunk 7]
```diff
@@ -41,10 +42,14 @@ export class Document {
     };
   }
 
+  getObjectId(): string {
```
Comment (lukasholzer) [inline packages/sketch-file-format/src/models/document.ts]: why we need this function?
you can directly access the objectID if you make it a public property

File: packages/sketch-file-format/src/models/document.ts [hunk 8]
```diff
@@ -4,15 +4,16 @@ import { Page } from './page';
...
+  private _objectID: string;
 
-  constructor(pages: Page[]) {
+  constructor(pages: Page[], objectID = UUID.generate()) {
     if (Document._instance) {
       return Document._instance;
     }
     Document._instance = this;
     this._pages = [...pages];
+    this._objectID = objectID;
```
Comment (lukasholzer) [inline packages/sketch-file-format/src/models/document.ts]: this.objectID = UUID.generate()

File: packages/sketch-builder/src/main.ts [hunk 9]
```diff
@@ -44,6 +54,11 @@ export async function main(config: SketchBuilderConfig, meta?: Library | undefin
   }
   const exitCode = await elementFetcher.generateSketchFile();
 
+  // Write (updated) objectIdMapping file.
```
Comment (lukasholzer) [inline packages/sketch-builder/src/main.ts]: can you explain why you check the config objectID mapping and the elementFetcher objectIDMapping @lara-aigmueller ?

File: packages/sketch-builder/src/helpers/object-id-handler.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Sets the object_ID of the given symbol.
+ * @param symbolName – The symbol's name, e.g. "button/light-bg/main/primary/default".
+ * @param objectId - The newly generated object_ID.
+ * @param objectIdMapping – The idMapping object.
+ * @returns The updated idMapping object.
+ */
+export function setObjectId(symbolName: string, objectId: string, objectIdMapping: ObjectIdMapping): ObjectIdMapping {
+  const newObjectIdMapping = Object.assign({}, objectIdMapping);
```
Comment (lukasholzer) [inline packages/sketch-builder/src/helpers/object-id-handler.ts]: you know that this only clones one level – (you may use lodash for this!)

File: packages/sketch-builder/src/helpers/object-id-handler.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  };
+
+  return newObjectIdMapping;
+}
+
+/**
+ * @param objectIdMapping - The idMapping object.
+ * @returns the library ID when given in the idMapping object, undefined else.
+ */
+export function getLibraryId(objectIdMapping: ObjectIdMapping) {
```
Comment (lukasholzer) [inline packages/sketch-builder/src/helpers/object-id-handler.ts]: can you specify a return type if the function?

File: packages/app-builder/src/angular-app-shell/files/__directory__/src/app/utils/mutate-variants.ts (hunk 1 above)
Comment (lara-aigmueller) [reply packages/app-builder/src/angular-app-shell/files/__directory__/src/app/utils/mutate-variants.ts]: created a ticket for the next sprint: UX-9384

File: packages/sketch-object-id-collector/src/main.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    await result.forEach((jsonFile) => {
+      const content = jsonFile.buffer.toString();
+      if (jsonFile.path === 'document.json') {
+        document = content;
+      } else {
+        pages.push(content);
+      }
+    });
+
+    const objectIdCollector = new ObjectIdCollector(document, pages);
```
Comment (lukasholzer) [inline packages/sketch-object-id-collector/src/main.ts]:
```typescript
let sketchDocument: SketchDocument;
let sketchPages: SketchPage[];

try {
  sketchDocument = JSON.parse(documnet);
  sketchPages = pages.map(page => JSON.parse(page));
} catch {
  throw new Error(`Parsing the Sketchfile failed`);
}

const collector = new ObjectIdCollector(sketchDocument);
collector.collect(sketchPages);
```

File: packages/sketch-object-id-collector/src/interfaces/object-id-mapping.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  version: string;
+  symbols: { [key: string]: ObjectIdMappingSymbol };
+  symbolAliases?: { [key: string]: string };
+  libraryId?: string;
+}
+
+export interface ObjectIdMappingSymbol {
+  objectId: string;
+  changeIdentifier: number;
+  overrides?: ObjectOverrideProperties[];
```
Comment (lukasholzer) [inline packages/sketch-object-id-collector/src/interfaces/object-id-mapping.ts]: make it not optional – default empty array *(is way more easy to check if you always have the property)*
maybe an object would be faster to handle where the id is the key…

File: packages/sketch-object-id-collector/src/interfaces/object-id-mapping.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+export interface ObjectIdMapping {
+  version: string;
+  symbols: { [key: string]: ObjectIdMappingSymbol };
+  symbolAliases?: { [key: string]: string };
+  libraryId?: string;
+}
+
+export interface ObjectIdMappingSymbol {
+  objectId: string;
```
Comment (lukasholzer) [inline packages/sketch-object-id-collector/src/interfaces/object-id-mapping.ts]: why you need the ID here if it is the key?

File: packages/sketch-object-id-collector/src/interfaces/object-id-mapping.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  changeIdentifier: number;
+  overrides?: ObjectOverrideProperties[];
+}
+
+export interface ObjectOverrideProperties {
+  objectId: string;
+  className: string;
```
Comment (lukasholzer) [inline packages/sketch-object-id-collector/src/interfaces/object-id-mapping.ts]: is the className the text?

File: packages/sketch-file-format/src/models/document.ts [hunk 16]
```diff
@@ -4,15 +4,16 @@ import { Page } from './page';
 
 export class Document {
   private static _instance: Document;
-  private static _objectID = UUID.generate();
   private _pages = [];
+  objectID: string;
```
Comment (lukasholzer) [inline packages/sketch-file-format/src/models/document.ts]: `objectID = UUID.generate();` and if you have to overwrite the id just reassign objectID from outside

```typescript
const doc = new Document([]);
doc.objectID = 'NEW_ID';
...
doc.generateObject()
```

File: packages/sketch-file-builder/src/sketch.ts [hunk 17]
```diff
@@ -43,6 +54,21 @@ export class Sketch {
     createDir(join(Sketch.TMP_PATH, 'images'));
   }
 
+  /**
+   * Generates a new document and sets the library ID when already defined.
+   * @param pages – Sketch pages.
+   * @returns The generated document.
+   */
+  private generateDocument(pages: Page[]): Document {
+    log.debug(chalk`\n\n\t{yellow ——— GENERATING DOCUMENT ———}\n`);
```
Comment (lukasholzer) [inline packages/sketch-file-builder/src/sketch.ts]: create one document with the pages

and if you have a libraryid the overwrite it

Comment (lukasholzer) [reply packages/sketch-file-builder/src/sketch.ts]: and why do you need a own function for that do it in the write function

File: packages/sketch-builder/src/object-id-service.ts [hunk 18]
```diff
@@ -0,0 +1,78 @@
+import { ObjectIdMapping } from './interfaces';
+import { IBounding, Text } from '@sketchmine/sketch-file-format';
+
+/**
+ * The object Id service holds an objectIdMapping object during the library build.
+ * The sketch builder uses this collection of object Ids as reference and reuses
+ * IDs when they have already been generated on a previous library build.
+ */
+export class ObjectIdService {
+  private static _instance: ObjectIdService;
```
Comment (lukasholzer) [inline packages/sketch-builder/src/object-id-service.ts]: please avoid underscores (legacy variant)

Comment (lukasholzer) [reply packages/sketch-builder/src/object-id-service.ts]: I know we have it multiple times with underscore but every time you see something with underscore please try to refactor it.

File: packages/sketch-builder/src/object-id-service.ts [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * Gets the objectID of the given symbol.
+   * @param symbolKey – The symbol's name, e.g. "button/light-bg/main/primary/default".
+   * @returns The symbol's objectID or undefined if there is none.
+   */
+  getObjectId(symbolKey: string): string | undefined {
+    let objectId;
+    if (this.collection && this.collection.symbols && this.collection.symbols[symbolKey]) {
+      objectId = this.collection.symbols[symbolKey].objectId;
```
Comment (lukasholzer) [inline packages/sketch-builder/src/object-id-service.ts]: why you don't return here?

File: packages/sketch-builder/src/object-id-service.ts [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * Returns the library ID of the objectID collection.
+   * @returns the library ID when given in the collection object, undefined else.
+   */
+  getLibraryId(): string | undefined {
+    if (this.collection && this.collection.libraryId) {
+      return this.collection.libraryId;
+    }
+
+    return;
```
Comment (lukasholzer) [inline packages/sketch-builder/src/object-id-service.ts]: not needed

Comment (lukasholzer) [reply packages/sketch-builder/src/object-id-service.ts]: getter would be better (faster) than a function execution

File: packages/sketch-builder/src/interfaces/object-id-mapping.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+import { IBounding } from '@sketchmine/sketch-file-format';
```
Comment (lukasholzer) [inline packages/sketch-builder/src/interfaces/object-id-mapping.ts]: can you import this interface from the new package and please remove that file

File: packages/sketch-builder/src/interfaces/index.ts [hunk 22]
```diff
@@ -0,0 +1 @@
+export * from './object-id-mapping';
```
Comment (lukasholzer) [inline packages/sketch-builder/src/interfaces/index.ts]: remove file

**Comment by lukasholzer** [comment]
We need some tests for:

[ ] creating everything new when no existing library file was provided
[ ] using the collected ids from an existing library file
[ ] adding a new component with new id to the existing file
[ ] removing an id when the component was deleted

**Comment by lara-aigmueller** [comment]
For the last two tests you mentioned (adding/removing a component before the next library build) we have to use memfs as you suggested so that we can modify the html content during the running test.

**Pull request merged**